### ISSUE TITLE
TF2rc Simple model subclass with Gradient tape error when transformed with autograph

### ISSUE THREAD

**Issue by fgr1986** [tensorflow/tensorflow#32656]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): https://colab.research.google.com/github/tensorflow/docs/blob/r2.0rc/site/en/r2/tutorials/quickstart/advanced.ipynb
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): TF2RC binary
- TensorFlow version (use command below): TF2RC
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
A simple toy example using Gradient tape and a subclass-model cannot be transformed into a graph with Autograph.
Error: _**could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4**_

**Describe the expected behavior**
The use of `@tf.function` should correctly create Autograph

**Code to reproduce the issue**
Please, run 
https://colab.research.google.com/github/tensorflow/docs/blob/r2.0rc/site/en/r2/tutorials/quickstart/advanced.ipynb

> 
> try:
>   # %tensorflow_version only exists in Colab.
>   %tensorflow_version 2.x
> except Exception:
>   pass
> from __future__ import absolute_import, division, print_function, unicode_literals
> 
> import tensorflow as tf
> 
> from tensorflow.keras.layers import Dense, Flatten, Conv2D
> from tensorflow.keras import Model
> mnist = tf.keras.datasets.mnist
> 
> (x_train, y_train), (x_test, y_test) = mnist.load_data()
> x_train, x_test = x_train / 255.0, x_test / 255.0
> 
> # Add a channels dimension
> x_train = x_train[..., tf.newaxis]
> x_test = x_test[..., tf.newaxis]
> train_ds = tf.data.Dataset.from_tensor_slices(
>     (x_train, y_train)).shuffle(10000).batch(32)
> 
> test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(32)
> class MyModel(Model):
>   def __init__(self):
>     super(MyModel, self).__init__()
>     self.conv1 = Conv2D(32, 3, activation='relu')
>     self.flatten = Flatten()
>     self.d1 = Dense(128, activation='relu')
>     self.d2 = Dense(10, activation='softmax')
> 
>   def call(self, x):
>     x = self.conv1(x)
>     x = self.flatten(x)
>     x = self.d1(x)
>     return self.d2(x)
> 
> # Create an instance of the model
> model = MyModel()
> loss_object = tf.keras.losses.SparseCategoricalCrossentropy()
> 
> optimizer = tf.keras.optimizers.Adam()
> train_loss = tf.keras.metrics.Mean(name='train_loss')
> train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='train_accuracy')
> 
> test_loss = tf.keras.metrics.Mean(name='test_loss')
> test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='test_accuracy')
> @tf.function
> def train_step(images, labels):
>   with tf.GradientTape() as tape:
>     predictions = model(images)
>     loss = loss_object(labels, predictions)
>   gradients = tape.gradient(loss, model.trainable_variables)
>   optimizer.apply_gradients(zip(gradients, model.trainable_variables))
> 
>   train_loss(loss)
>   train_accuracy(labels, predictions)
> @tf.function
> def test_step(images, labels):
>   predictions = model(images)
>   t_loss = loss_object(labels, predictions)
> 
>   test_loss(t_loss)
>   test_accuracy(labels, predictions)
> EPOCHS = 5
> 
> for epoch in range(EPOCHS):
>   for images, labels in train_ds:
>     train_step(images, labels)
> 
>   for test_images, test_labels in test_ds:
>     test_step(test_images, test_labels)
> 
>   template = 'Epoch {}, Loss: {}, Accuracy: {}, Test Loss: {}, Test Accuracy: {}'
>   print(template.format(epoch+1,
>                         train_loss.result(),
>                         train_accuracy.result()*100,
>                         test_loss.result(),
>                         test_accuracy.result()*100))
> 
>   # Reset the metrics for the next epoch
>   train_loss.reset_states()
>   train_accuracy.reset_states()
>   test_loss.reset_states()
>   test_accuracy.reset_states()

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by oanush** [comment]
@fgr1986 ,
Given code worked with TF-2.0rc1 without any issue's, please take a look at the [gist](https://github.com/tensorflow/tensorflow/issues/32651) of colab. Try running the code in TF-2.0rc1

**Comment by fgr1986** [comment]
Hi @oanush , just reproduced the problem with ' 2.0.0-rc1'. Please, note that the error is a warning, (it continues execution but it does not correctly pass the autograph function.

You can see the issue putting ' print(tf.__version__)' in https://colab.research.google.com/github/tensorflow/docs/blob/r2.0rc/site/en/r2/tutorials/quickstart/advanced.ipynb

**Comment by oanush** [comment]
@fgr1986 ,
I tried again and I did not face and error for TF-2.0rc1. Let us know exactly what warnings are received.

**Comment by fgr1986** [comment]
Hi @oanush .
I do not know where the problem to reproduce the issue may be.
1) Right now, I just opened: https://colab.research.google.com/github/tensorflow/docs/blob/r2.0rc/site/en/r2/tutorials/quickstart/advanced.ipynb
Added `import tensorflow as tf
print(tf.__version__)`
Which reports version `2.0.0-rc1`

2) Changed the runtime to GPU

3) Run all, and get the following warnings:
```
WARNING:tensorflow:Entity <function train_step at 0x7f47c59bb268> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Str'
WARNING: Entity <function train_step at 0x7f47c59bb268> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Str'
WARNING:tensorflow:Layer my_model is casting an input tensor from dtype float64 to the layer's dtype of float32, which is new behavior in TensorFlow 2.  The layer has dtype float32 because it's dtype defaults to floatx.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

WARNING:tensorflow:Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING: Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7f47c4994c80> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
WARNING: Entity <function Function._initialize_uninitialized_variables.<locals>.initialize_variables at 0x7f47c4994c80> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Num'
WARNING:tensorflow:Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING: Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function test_step at 0x7f47c59bb9d8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Str'
WARNING: Entity <function test_step at 0x7f47c59bb9d8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: module 'gast' has no attribute 'Str'
WARNING:tensorflow:Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING: Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING: Entity <bound method MyModel.call of <__main__.MyModel object at 0x7f47c59f7390>> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4

```

Thanks

**Comment by jvishnuvardhan** [comment]
@fgr1986 I ran it in `2.0.0rc2` with GPU (as mentioned by you) and it runs without any issue. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/a2ee547861e319f66974c23e9c288927/advanced.ipynb). 

Here is the error trace
```
WARNING:tensorflow:Layer my_model is casting an input tensor from dtype float64 to the layer's dtype of float32, which is new behavior in TensorFlow 2.  The layer has dtype float32 because it's dtype defaults to floatx.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

Epoch 1, Loss: 0.13170050084590912, Accuracy: 95.97166442871094, Test Loss: 0.06416505575180054, Test Accuracy: 97.98999786376953
Epoch 2, Loss: 0.04071308672428131, Accuracy: 98.74166870117188, Test Loss: 0.05933062732219696, Test Accuracy: 98.15999603271484
Epoch 3, Loss: 0.022074595093727112, Accuracy: 99.30333709716797, Test Loss: 0.053413353860378265, Test Accuracy: 98.33999633789062
Epoch 4, Loss: 0.013294064439833164, Accuracy: 99.55500030517578, Test Loss: 0.05290050432085991, Test Accuracy: 98.40999603271484
Epoch 5, Loss: 0.008136439137160778, Accuracy: 99.73500061035156, Test Loss: 0.06418339163064957, Test Accuracy: 98.2699966430664
```

If you set the default `dtype`  as `tf.keras.backend.set_floatx('float64')`, then the output is 

```
Epoch 1, Loss: 0.14186638195651272, Accuracy: 95.70166666666667, Test Loss: 0.05724755003185461, Test Accuracy: 98.11999999999999
Epoch 2, Loss: 0.043617788805440066, Accuracy: 98.65666666666667, Test Loss: 0.047525364846504776, Test Accuracy: 98.49
Epoch 3, Loss: 0.022846894193099192, Accuracy: 99.23833333333333, Test Loss: 0.0562137230045946, Test Accuracy: 98.22
Epoch 4, Loss: 0.01501299387764108, Accuracy: 99.515, Test Loss: 0.05169376732314508, Test Accuracy: 98.38
Epoch 5, Loss: 0.008679648822207916, Accuracy: 99.715, Test Loss: 0.062491611017893994, Test Accuracy: 98.33
```

Please close the issue if it was resolved by `2.0.0rc2`. Thanks!

**Comment by fgr1986** [comment]
Yeap, rc2 worked nicely. 
Thanks
